spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -219,6 +201,45 @@ trait DeltaTypeWideningAlterTableTests extends QueryErrorsBase {
       DecimalType(Decimal.MAX_INT_DIGITS + 3, 1),
       Seq(BigDecimal("-67.89"), BigDecimal("9" * (Decimal.MAX_LONG_DIGITS - 2) + ".99")))
   )
+}
+
+/**
+ * Trait collecting a subset of tests providing core coverage for type widening using ALTER TABLE
+ * CHANGE COLUMN TYPE.
+ */
+trait DeltaTypeWideningAlterTableTests extends QueryErrorsBase with DeltaTypeWideningTestCases {
+  self: QueryTest with ParquetTest with DeltaTypeWideningTestMixin with DeltaDMLTestUtils =>
```

Comment:
I don't understand all the usecases for this type of reverse inheritance, but could these just be normal mixins or there is a deeper reason?

Reply:
I use it mainly as a way to document dependencies on classes/traits: putting in the self type dependencies to test helpers/classes that are typically shared with other tests and directly inheriting traits that are specific to this test trait.

There's no functional reason behind it, if that's hurts readability I'll stop doing it.

Reply:
Yeah, it makes sense, you are basically putting limitations on which suites/traits are allowed to inherit from this one. 
One more a less legit use case I've seen is us breaking up a large MergeInto class into several traits and each trait also using the methods of the "child" class. Basically just organizing the code better without having to pass around too much state explicitly in constructor parameters or method args.